matchtree.go

```diff
@@ -411,18 +411,55 @@ func (t *regexpMatchTree) matches(cp *contentProvider, cost int, known map[match
 	idxs := t.regexp.FindAllIndex(cp.data(t.fileName), -1)
 	found := t.found[:0]
 	for _, idx := range idxs {
-		found = append(found, &candidateMatch{
+		cm := &candidateMatch{
 			byteOffset:  uint32(idx[0]),
 			byteMatchSz: uint32(idx[1] - idx[0]),
 			fileName:    t.fileName,
-		})
+		}
+
+		found = append(found, cm)
 	}
 	t.found = found
 	t.reEvaluated = true
 
 	return len(t.found) > 0, true
 }
 
+// breakMatchesOnNewlines returns matches resulting from breaking each element
+// of cms on newlines within text.
+func breakMatchesOnNewlines(cms []*candidateMatch, text []byte) []*candidateMatch {
+	var cms2 []*candidateMatch
+	for _, cm := range cms {
+		cms2 = append(cms2, breakOnNewlines(cm, text)...)
+	}
+	return cms2
+}
+
+// breakOnNewlines returns matches resulting from breaking cm on newlines
+// within text.
+func breakOnNewlines(cm *candidateMatch, text []byte) []*candidateMatch {
+	var cms []*candidateMatch
+	var addMe = &candidateMatch{}
+	*addMe = *cm
+	for i := uint32(cm.byteOffset); i < cm.byteOffset+cm.byteMatchSz; i++ {
+		if text[i] == '\n' {
+			addMe.byteMatchSz = i - addMe.byteOffset
```

Comment:
```suggestion
			addMe.byteMatchSz = i - cm.byteOffset
```

`addMe.byteOffset` is just `cm.byteOffset` not some state being mutated as part of `addMe` so I think this is more clear.

Reply:
I stand corrected, it is state being mutated on line 453 I now I understand